keps/sig-node/4563-eviction-request-api/README.md

```diff
@@ -0,0 +1,2163 @@
+recommended to check `creationTimestamp` of the EvictionRequests and observe
+`evictionrequest_controller_active_interceptor` metric to see how much it takes for an interceptor
+to complete the eviction. This metric can be also used to implement additional alerts.
+
+#### Disruptive Eviction
+
+When using kubectl drain, pods without owning controller and pods with local storage
+(having `emptyDir` volumes) are not evicted by default. We have decided to evict most
+of the pods (except DaemonSet and mirror pods) by default. In the motivation section of the
+[Declarative Node Maintenance KEP](https://github.com/kubernetes/enhancements/pull/4213),
+we can see many administrators override these default settings and many components evict all pods
+indiscriminately. There are also many ways that users use to prevent the eviction;
+PodDisruptionBudgets, validating admission webhooks, or just plain HA. Users who want to protect
+their applications in today's clusters should already be aware that they should be able to handle
+sudden evictions. Therefore, it should be okay for the eviction request controller to evict these
+pods.
+
+To mitigate the sudden eviction problem, users should use PodDisruptionBudgets or HA.
+
+## Design Details
+
+### EvictionRequest
+
+We will introduce a new type called `evictionrequest.coordination.k8s.io` to enforce the contract
+between the eviction requesters and the interceptors. This type is a bit similar to
+`leases.coordination.k8s.io` in that it requires multiple actors to synchronize the state. Which in
+our case is the progress of the eviction.
+
+### Eviction Requester
+[Eviction Requester](#eviction-requester) section provides a general overview.
+
+There can be many eviction requesters for a single EvictionRequest.
+
+When a requester decides that a pod needs to be evicted, it should create an EvictionRequest:
+- `.metadata.generateName` should be set instead of the `.metadata.name` to avoid conflicts.
+  Conflicts could occur due to pods with the same name but a different UIDs. In the future, it might
+  be possible that EvictionRequest will be extended for other resources than pods (e.g. PVCs), which
+  could result in name conflicts. For more details, see [The Name of the EvictionRequest Objects](#the-name-of-the-evictionrequest-objects)
+  alternatives section.
+- `.spec.type` should be set to `Soft` (default value) since it is currently the only supported type.
+- `.spec.target.podRef` should be set to fully identify the pod. The name and the UID should be
+  specified to ensure that we do not evict a pod with the same name that appears immediately
+  after the previous pod is removed.
+- `.spec.heartbeatDeadlineSeconds` should be set to a reasonable value. It is recommended to leave
+  it at the default value of 1800 (30m) to allow for potential interceptor disruptions.
+- `.spec.requesters` It should add itself (requester subdomain) to the requesters list upon creation.
```

Comment:
Assume multiple things want to evict the same pod.  Do we get multiple `EvictionRequest` objects with a single `requester` each, or do we get a single `EvictionRequest` with multiple `requesters` ?